pkg/controller/installation/products/mobiledeveloperconsole/reconciler_test.go

```diff
@@ -0,0 +1,583 @@
+			Installation:   &v1alpha1.Installation{},
+			FakeClient:     fakeclient.NewFakeClient(),
+			FakeConfig: &config.ConfigReadWriterMock{
+				ReadMobileDeveloperConsoleFunc: func() (ready *config.MobileDeveloperConsole, e error) {
+					return nil, errors.New("could not read mdc config")
+				},
+			},
+		},
+		{
+			Name:         "test namespace is set without fail",
+			Installation: &v1alpha1.Installation{},
+			FakeClient:   fakeclient.NewFakeClient(),
+			FakeConfig: &config.ConfigReadWriterMock{
+				ReadMobileDeveloperConsoleFunc: func() (ready *config.MobileDeveloperConsole, e error) {
+					return config.NewMobileDeveloperConsole(config.ProductConfig{
+						"NAMESPACE": "",
+					}), nil
+				},
+			},
+		},
+		{
+			Name:           "test subscription phase with error from mpm",
+			ExpectedStatus: v1alpha1.PhaseFailed,
+			ExpectError:    true,
+			Installation:   &v1alpha1.Installation{},
+			FakeMPM: &marketplace.MarketplaceInterfaceMock{
+				InstallOperatorFunc: func(ctx context.Context, serverClient pkgclient.Client, owner ownerutil.Owner, os marketplacev1.OperatorSource, t marketplace.Target, operatorGroupNamespaces []string, approvalStrategy operatorsv1alpha1.Approval) error {
+					return errors.New("dummy error")
+				},
+			},
+			FakeClient: fakeclient.NewFakeClient(),
+			FakeConfig: basicConfigMock(),
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.Name, func(t *testing.T) {
+			_, err := NewReconciler(tc.FakeConfig, tc.Installation, tc.FakeMPM)
+			if err != nil && err.Error() != tc.ExpectedError {
+				t.Fatalf("unexpected error : '%v', expected: '%v'", err, tc.ExpectedError)
+			}
+			if err == nil && tc.ExpectedError != "" {
+				t.Fatalf("expected error '%v' and got nil", tc.ExpectedError)
+			}
+		})
+	}
+}
+
+func TestReconciler_reconcileCustomResource(t *testing.T) {
+	scheme, err := getBuildScheme()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cases := []struct {
+		Name           string
+		FakeClient     pkgclient.Client
+		FakeConfig     *config.ConfigReadWriterMock
+		Installation   *v1alpha1.Installation
+		ExpectError    bool
+		ExpectedStatus v1alpha1.StatusPhase
+		FakeMPM        *marketplace.MarketplaceInterfaceMock
+	}{
+		{
+			Name:           "Test reconcile custom resource returns in progress when successful created",
+			FakeClient:     fakeclient.NewFakeClientWithScheme(scheme, getOauthClientSecret(), getOperatorDC(), getRoute()),
+			FakeConfig:     basicConfigMock(),
+			Installation:   &v1alpha1.Installation{},
+			ExpectedStatus: v1alpha1.PhaseCompleted,
+		},
+		{
+			Name: "Test reconcile custom resource returns failed when cr status is failed",
+			FakeClient: &moqclient.SigsClientInterfaceMock{
+				CreateFunc: func(ctx context.Context, obj runtime.Object) error {
+					return errors.New("dummy create error")
+				},
+				GetFunc: func(ctx context.Context, key types.NamespacedName, obj runtime.Object) error {
+					return errors.New("get request failed")
+				},
+			},
```

Comment:
Can probably return a slightly better error like the one [here](https://github.com/integr8ly/integreatly-operator/blob/master/pkg/controller/installation/products/mobilesecurityservice/reconciler_test.go#L160-L165)